src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
     return genericType instanceof Class ? (Class<?>) genericType : Unknown.class;
   }
 
+  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableTypeMap) {
+    // Check for terminal cases to avoid allocating HashMap and trivial call.
+    if (genericType == null)
+      return null;
+    else if (genericType instanceof Class<?>)
+      return genericType;
+    else
+      return reify(genericType, typeVariableTypeMap, new HashMap<>());
+  }
+
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
+    // Terminal cases.
+    if (genericType == null)
       return null;
-    } else if (genericType instanceof Class<?>) {
+    else if (genericType instanceof Class<?>)
       return genericType;
-    } else if (genericType instanceof ParameterizedType) {
+    else if (cache.containsKey(genericType)) {
+      while (cache.containsKey(genericType)) {
```

Comment:
Can you describe when the while loop becomes necessary?

Reply:
For when the cycle inside the type structure is of a length greater than one, it must be traversed.